client/shared/src/test/scala/org/http4s/client/ClientSuite.scala

```diff
@@ -147,4 +150,104 @@ class ClientSpec extends Http4sSuite with Http4sDsl[IO] {
       .attempt
       .assertEquals(Left(MyThrowable), "Throwable did not get handled as expected")
   }
+
+  test("mock client should drain the body if it has not been consumed") {
+
+    def app(compiled: Ref[IO, Int]) = HttpApp[IO] { (_: Request[IO]) =>
+      Response[IO]()
+        .pipeBodyThrough(_.onFinalize(compiled.update(_ + 1)))
+        .pure[IO]
+    }
+
+    for {
+      compiledCounter <- Ref.of[IO, Int](0)
+      client = Client.fromHttpApp(app(compiledCounter))
+      _ <- client.status(Request[IO]()).void
+      compiled <- compiledCounter.get
+    } yield assertEquals(compiled, 1)
+  }
+
+  test("mock client should drain the body if it has been partially consumed") {
+
+    val entity = asciiBytes"foo"
+
+    def app(channel: Channel[IO, Byte], compiled: Ref[IO, Int]) = HttpApp[IO] { (_: Request[IO]) =>
+      Response[IO]()
+        .withEntity(entity)
+        .pipeBodyThrough(_.evalTap(channel.send).onFinalize(compiled.update(_ + 1)))
+        .pure[IO]
+    }
+
+    for {
+      compiledCounter <- Ref.of[IO, Int](0)
+      channel <- Channel.unbounded[IO, Byte]
+      client = Client.fromHttpApp(app(channel, compiledCounter))
+      result <- client.run(Request[IO]()).use(_.body.take(1).compile.toVector)
+      _ <- channel.close.void
+      resultWithDrained <- channel.stream.compile.toVector
+      compiled <- compiledCounter.get
+    } yield {
+      assertEquals(result, entity.take(1).toSeq.toVector)
+      assertEquals(resultWithDrained, entity.toSeq.toVector)
+      assertEquals(compiled, 1)
+    }
+
+  }
+
+  test("mock client should fail if the internal channel is not synchronous") {
+
+    def app(compiled: Ref[IO, Int]) = HttpApp[IO] { (_: Request[IO]) =>
+      Response[IO]()
+        .withEntity("foo")
+        .pipeBodyThrough(_.onFinalize(compiled.update(_ + 1)))
+        .pure[IO]
+    }
+
+    for {
+      compiledCounter <- Ref.of[IO, Int](0)
+      client = Client.fromHttpApp(app(compiledCounter))
+      _ <- client.run(Request[IO]()).use { _ =>
+        IO.sleep(1.millis) *>
+          compiledCounter.get.assertEquals(0)
+      }
+      _ <- compiledCounter.get.assertEquals(1)
+    } yield ()
```

Comment:
We can use `executeEmbed` for reliable, time-based tests, since it uses a simulated clock :)

```suggestion
    val test = for {
      compiledCounter <- Ref.of[IO, Int](0)
      client = Client.fromHttpApp(app(compiledCounter))
      _ <- client.run(Request[IO]()).use { _ =>
        IO.sleep(1.second) *>
          compiledCounter.get.assertEquals(0)
      }
      _ <- compiledCounter.get.assertEquals(1)
    } yield ()
    
    TestControl.executeEmbed(test)
```